lib/StripeObject.php

```diff
@@ -205,81 +166,170 @@ public static function constructFrom($values, $opts)
-                if ($v === null) {
-                    $v = '';
-                }
+        $updateParams = [];
 
-                $params[$k] = $v;
+        foreach ($this->_values as $k => $v) {
+            // There are a few reasons that we may want to add in a parameter for
+            // update:
+            //
+            //   1. The `$force` option has been set.
+            //   2. We know that it was modified.
+            //   3. Its value is a StripeObject. A StripeObject may contain modified
+            //      values within in that its parent StripeObject doesn't know about.
+            //
+            $original = array_key_exists($k, $this->_originalValues) ? $this->_originalValues[$k] : null;
+            $unsaved = $this->_unsavedValues->includes($k);
+            if ($force || $unsaved || $v instanceof StripeObject) {
+                $updateParams[$k] = $this->serializeParamsValue(
+                    $this->_values[$k],
+                    $original,
+                    $unsaved,
+                    $force,
+                    $k
+                );
             }
         }
 
-        // Get nested updates.
-        foreach (self::$nestedUpdatableAttributes->toArray() as $property) {
-            if (isset($this->$property)) {
-                if ($this->$property instanceof StripeObject) {
-                    $serialized = $this->$property->serializeParameters();
-                    if ($serialized) {
-                        $params[$property] = $serialized;
-                    }
+        // a `null` that makes it out of `serializeParamsValue` signals an empty
+        // value that we shouldn't appear in the serialized form of the object
+        $updateParams = array_filter(
+            $updateParams,
+            function ($v) {
+                return $v !== null;
+            }
+        );
+
+        return $updateParams;
+    }
+
+
+    public function serializeParamsValue($value, $original, $unsaved, $force, $key = null)
+    {
+        // The logic here is that essentially any object embedded in another
+        // object that had a `type` is actually an API resource of a different
+        // type that's been included in the response. These other resources must
+        // be updated from their proper endpoints, and therefore they are not
+        // included when serializing even if they've been modified.
+        //
+        // There are _some_ known exceptions though.
+        //
+        // For example, if the value is unsaved (meaning the user has set it), and
+        // it looks like the API resource is persisted with an ID, then we include
+        // the object so that parameters are serialized with a reference to its
+        // ID.
+        //
+        // Another example is that on save API calls it's sometimes desirable to
+        // update a customer's default source by setting a new card (or other)
+        // object with `#source=` and then saving the customer. The
```

Comment:
The `#` designation for instance variables is fairly Ruby specific. Is there a more conventional way for PHP? (Maybe just `.`.)

Reply:
I used `->source =`. I'm not sure how conventional it is but I think it's fairly explicit.